src/node.h

```diff
@@ -353,18 +354,88 @@ NODE_EXTERN IsolateData* CreateIsolateData(
     ArrayBufferAllocator* allocator = nullptr);
 NODE_EXTERN void FreeIsolateData(IsolateData* isolate_data);
 
-// TODO(addaleax): Add an official variant using STL containers, and move
-// per-Environment options parsing here.
+struct ThreadId {
+  uint64_t id = static_cast<uint64_t>(-1);
+};
+NODE_EXTERN ThreadId AllocateEnvironmentThreadId();
+
+namespace EnvironmentFlags {
+enum Flags : uint64_t {
+  kNoFlags = 0,
+  // Use the default behaviour for Node.js instances.
+  kDefaultFlags = 1 << 0,
+  // Controls whether this Environment is allowed to affect per-process state
+  // (e.g. cwd, process title, uid, etc.).
+  // This is set when using kDefaultFlags.
+  kOwnsProcessState = 1 << 1,
+  // Set if this Environment instance is associated with the global inspector
+  // handling code (i.e. listening on SIGUSR1).
+  // This is set when using kDefaultFlags.
+  kOwnsInspector = 1 << 2,
+  // Set up a standard Node.js Environment, including starting any libuv handles
+  // and the inspector, as well as anything dependent on e.g. environment
+  // variables or command line flags.
+  // This conflicts with LoadEnvironment().
+  // It is recommended to not set or rely on this flag, and it will be removed.
+  kPrepareForExecution = 1 << 3,
```

Comment:
This currently runs the preparation for main thread execution - which means creating two environments with this flag is going to be kind of wonky.

Is this the best way to divide the options? It also seems possible to do this preparation step in a method and the user can choose to call it themselves. The flags here seem to anchor on the decision that Environment should be the primary abstraction for embedders to interact with the Node.js instance, but I believe previously it is more like an opaque structure being passed around (albeit being hacked a lot in the wild)?

My original plan is to make an abstraction over Worker and NodeMainInstance somehow, and let the users interact with the instance primarily through that. Then Environment would become more of an implementation detail that we are free to refactor with. AFAICT, the embedding use cases of Node.js mostly differ on whether the embdder wants to own the platform, or the event loop, etc. and I think making options out of those needs first and add a few more setters for lower-level options when they are not flexible enough instead of exposing low-level options from the start would result in a more maintainable API.

Reply:
> This currently runs the preparation for main thread execution - which means creating two environments with this flag is going to be kind of wonky.

Just to be clear, *this* particular flag should go away anyway. It’s only here because `CreateEnvironment()` does run `prepareMainThreadExecution()` currently.

> Is this the best way to divide the options? It also seems possible to do this preparation step in a method and the user can choose to call it themselves.

If you’re suggesting to add more flags that control which parts of the setup to run, then I think that’s a good idea. What flags would you have in mind?

> The flags here seem to anchor on the decision that Environment should be the primary abstraction for embedders to interact with the Node.js instance, but I believe previously it is more like an opaque structure being passed around (albeit being hacked a lot in the wild)?

I would say that `Environment` *is* the Node.js instance, so it should be what embedders interact with when they’re dealing with Node.js instances, too.

> My original plan is to make an abstraction over Worker and NodeMainInstance somehow, and let the users interact with the instance primarily through that. Then Environment would become more of an implementation detail that we are free to refactor with. AFAICT, the embedding use cases of Node.js mostly differ on whether the embdder wants to own the platform, or the event loop, etc. and I think making options out of those needs first and add a few more setters for lower-level options when they are not flexible enough instead of exposing low-level options from the start would result in a more maintainable API.

I’m sorry, but I disagree on this… `Environment` is not an implementation detail to me, and embedders really *should* have fine-grained control over the full setup of Platform/Isolate/event loop if they want to.

There can always be a helper that abstracts away things like V8/libuv setup, but it’s imo better to expose the more powerful API first rather than restricting usage to only specific patterns.

> That would also help us add more cctests for different scenarios without sneaking into the internal namespace.

I don’t really see how that makes it easier to avoid the internal namespace than the approa3h here? This API should give an embedder everything they need to at least put a Node.js instance into a usable state, and the same goes for our cctests.

Reply:
@joyeecheung Maybe it's better to put it this way... If `Environment` is an implementation detail, then what would you have in mind in terms of public API for managing Node.js instances and how does that differ from `Environment`? I think that would clear things up for me a bit...

Reply:
I'd definitely have to agree that `Environment` is not an internal implementation detail. I'm generally not all that happy with the way `Environment` is defined in terms of "public" api but it's what we've got.